### ISSUE TITLE
[Question] Plotting Continued Models on the Same Line

### ISSUE THREAD

**Issue by Zachary-Fernandes** [DLR-RM/rl-baselines3-zoo#364]
### ❓ Question

Hello, I have a question regarding plotting in rl-baselines3-zoo. I work on a cluster that limits runs to at most six hours, so I thought it would be a good idea to use checkpoints to save my runs. After I ran out of time, I scheduled a new job and continued from rl_model_9000000_steps.zip for another million steps, and this ran as expected.

However, two things occurred. First, the continued run's contents went into a different directory from the original run (DemonAttack-v4_2 instead of DemonAttack-v4_1). Second, when I tried to plot it with plot_train.py, it treated these directories as different runs.

How can I combine these two runs into one? My hope is to make the second run extend the first run as intended with checkpoints. Below is the plot made by plot_train.py.

![image](https://user-images.githubusercontent.com/32844805/223773386-fa96b89e-d2bb-421c-b39c-90c13e2179af.png)

Also attached are the contents for both DemonAttack-v4 directories.

![image](https://user-images.githubusercontent.com/32844805/223773693-720fffb7-e028-43ed-a448-bbc2c9e1ab39.png)

![image](https://user-images.githubusercontent.com/32844805/223773763-ced6a000-4827-4ef4-b5eb-7028792c75b5.png)

Thank you in advance.

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/rl-baselines3-zoo/issues) in the repo
- [X] I have read the [SB3 documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] I have read the [RL Zoo README](https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/README.md)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
> How can I combine these two runs into one? My hope is to make the second run extend the first run as intended with checkpoints. 

For that you will need to combine the csv files and then put them in a single folder.
Or offset manually the second run in `plot_train.py`.

**Comment by Zachary-Fernandes** [comment]
Thank you for the reply! Will these methods work with `all_plots.py` as well? I just wanted to make sure for plotting training and evaluation.
Also, how would one manually offset the second run within the script?

**Comment by araffin** [comment]
> Will these methods work with all_plots.py as well? I just wanted to make sure for plotting training and evaluation.

For evaluation, you will need to merge the `evaluations.npz` files.

>  Also, how would one manually offset the second run within the script?

What did you try so far?

**Comment by Zachary-Fernandes** [comment]
I tried looking through the [plotting code](https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/rl_zoo3/plots/all_plots.py) where the npz files were used. I never used npz files before, but based on what I looked up, it appears `evaluations.npz` is a dictionary where each metric, such as "timesteps" and "results", is a key that leads to a list of values relating to that metric, such as timesteps or a list of rewards earned during an evaluation episode. If this is the case, merging these npz files should be similar to merging the csv files. What other properties do they have?

Speaking of which, I took your suggestion on combining the csv files and made a Google Colab notebook to perform this task. I tested it by merging three 100000-timestep runs of CartPole-v1 using A2C and PPO, and below is the end result. It works best with the default rolling window size. I will continue with this notebook for when I implementing the npz merging code.

![image](https://user-images.githubusercontent.com/32844805/224037415-c4f971dc-16ec-49e1-9d35-6afc553f7054.png)

**Comment by araffin** [comment]
> made a Google Colab notebook to perform this task. 

could you share the link as it might be useful for others?

> it appears evaluations.npz is a dictionary where each metric, such as "timesteps" and "results", is a key that leads to a list of values relating to that metri

yes, it's a dictionary of numpy arrays.

**Comment by Zachary-Fernandes** [comment]
Yes. I will admit this is a work in progress as I still need to implement npy merging. This was also made for the project I am working on, which sets up a fork of rl-baselines3-zoo.

[Here](https://colab.research.google.com/drive/1yxQtzo0arfBPTnaead9pDJ4ck6OdJ9Mu?usp=sharing) is the Google Colab link.

For evaluations.npz, is there a way to get all the keys?

**Comment by araffin** [comment]
> For evaluations.npz, is there a way to get all the keys?

`.keys()`?

**Comment by Zachary-Fernandes** [comment]
Thank you. I tried this and was curious why it would not print the keys, but I found a [Stack Overflow post](https://stackoverflow.com/questions/55014142/i-cannot-see-any-keys-returned-when-i-did-np-loadfilename-npz-then-keys) that stated `.keys()` returned an iterable. With a print statement, I can see the npz structure is similar to the csv. I should have it implemented later today.

**Comment by Zachary-Fernandes** [comment]
Following up to this thread, I have a question regarding saving checkpoints in rl-baselines3-zoo. Say I want to train an agent with Asterix-v4 for 40 million timesteps. Within the six-hour limit, training can run through 10 million timesteps. When it concludes, three model zip files stand out to me: `Asterix-v4.zip`, `best_model.zip`, and `rl_model_10000000_steps.zip`.

What are the differences between these model files? If I wanted to resume training for another 10 million timesteps (with the aim being to eventually reach 40 million time steps), which of these would be the best to use?

I wanted to ask because when I resumed training with `Asterix-v4.zip`, it seemed like all of the models' rewards dropped after the first stop point (10 million iterations) almost like it resumed from an earlier point. Attached is an image demonstrating this alongside another showing a similar drop in `Breakout-v4`. With this in mind, I was curious if this worse performance could be attributed to which file I used to resumed.

![image](https://user-images.githubusercontent.com/32844805/224472348-4c0573ad-1789-47a7-bd10-7a50b2fb2612.png)

![image](https://user-images.githubusercontent.com/32844805/224472354-482bf47d-d0da-44ce-bc9b-97d4e989ca74.png)

**Comment by araffin** [comment]
> What are the differences between these model files?

one is a checkpoint, this other is saved at the end of training, the last one is the best model according to the evaluation callback.

> which of these would be the best to use?

Usually the one saved at the end of training, but that's not always true (for instance, if there is a performance drop just before the end of training).

> t seemed like all of the models' rewards dropped after the first stop point 

That's probably due to the fact you are using schedules and schedule are reset when resuming training.
See https://github.com/DLR-RM/stable-baselines3/issues/435 and related issues.

**Comment by Zachary-Fernandes** [comment]
I see. When you refer to schedules resetting while resuming training, is this whether `reset_num_timesteps` is set to `True` or `False` in `model.learn()`?
